waldo/add.go

```diff
@@ -0,0 +1,237 @@
+package waldo
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/waldoapp/waldo-go-cli/lib"
+	"github.com/waldoapp/waldo-go-cli/waldo/data"
+	"github.com/waldoapp/waldo-go-cli/waldo/data/tool"
+)
+
+type AddOptions struct {
+	AppName     string
+	Platform    string
+	RecipeName  string
+	UploadToken string
+	Verbose     bool
+}
+
+type AddAction struct {
+	ioStreams    *lib.IOStreams
+	options      *AddOptions
+	promptReader *lib.PromptReader
+	runtimeInfo  *lib.RuntimeInfo
+}
+
+//-----------------------------------------------------------------------------
+
+func NewAddAction(options *AddOptions, ioStreams *lib.IOStreams) *AddAction {
+	runtimeInfo := lib.DetectRuntimeInfo()
+
+	return &AddAction{
+		ioStreams:    ioStreams,
+		options:      options,
+		promptReader: ioStreams.PromptReader(),
+		runtimeInfo:  runtimeInfo}
+}
+
+//-----------------------------------------------------------------------------
+
+func (aa *AddAction) Perform() error {
+	var (
+		results []*tool.FoundBuildPath
+		recipe  *data.Recipe
+		cfg     *data.Configuration
+		err     error
+	)
+
+	err = data.ValidateRecipeName(aa.options.RecipeName)
+
+	if err == nil && len(aa.options.UploadToken) > 0 {
```

Comment:
The succession of instructions is not really go-like. If you get an error at the beginning it will test all the ifs.

You should probably rewrite this in the traditional go way:
```
if err := data.ValidateRecipeName(aa.options.RecipeName); err != nil || len(aa.options.UploadToken) = 0 {
return err;
}
...
```